File: lib/mix/tasks/agent.setup.ex
defmodule Mix.Tasks.Agent.Setup do
  @moduledoc """
  Automates the isolation of an AI agent using Git Worktrees.
  One Agent = One Branch = One Directory = One Port.

  Usage: mix agent.setup <agent_name>
  """
  use Mix.Task
  require Logger

  @agents_root "agents"
  @base_port 4001

  @impl Mix.Task
  def run([name | _]) do
    agent_dir = Path.join(@agents_root, name)
    branch = "agent/#{name}"

    IO.puts("\n\e[1;36m--- 🤖 Setting up Sandbox for Agent: #{name} ---\e[0m")

    # 1. Create Branch & Worktree
    setup_worktree(agent_dir, branch)

    # 2. Assign Unique Port
    port = assign_port(name, agent_dir)

    # 3. Trust Mise configuration (Parity Protection)
    trust_mise(agent_dir)

    # 4. Symlink Heavy Dependencies (Parity and Space Protection)
    symlink_assets(agent_dir)

    # 4. Final Verification
    IO.puts("\n\e[1;32m✅ Sandbox Ready at: #{agent_dir}\e[0m")
    IO.puts("\e[1mAgent Branch:\e[0m #{branch}")
    IO.puts("\e[1mParallel Port:\e[0m #{port}")
    IO.puts("\e[2m(To start the agent's server, run: PORT=#{port} iex -S mix phx.server in that directory)\e[0m\n")
  end

  def run(_) do
    Mix.raise("Usage: mix agent.setup <agent_name>")
  end

  # --- Internal Logic ---

  defp setup_worktree(dir, branch) do
    IO.puts("📦 Creating Git Worktree...")

    # Ensure agent root exists (but ignore it from git)
    File.mkdir_p!(@agents_root)
    ensure_ignored()

    # Create branch if doesn't exist
    case System.cmd("git", ["rev-parse", "--verify", branch], stderr_to_stdout: true) do
      {_, 0} -> IO.puts("   (Using existing branch: #{branch})")
      _ -> System.cmd("git", ["branch", branch, "main"])
    end

    # Add Worktree
    case System.cmd("git", ["worktree", "add", dir, branch]) do
      {_, 0} -> :ok
      {msg, _} -> Mix.raise("Could not create worktree: #{msg}")
    end
  end

  defp assign_port(name, agent_dir) do
    # Scan existing agents for ports to avoid collisions
    existing_ports =
      Path.wildcard("#{@agents_root}/*/.env")
      |> Enum.map(fn path ->
        case File.read(path) do
          {:ok, content} ->
            case Regex.run(~r/PORT=(\d+)/, content) do
              [_, p] -> String.to_integer(p)
              _ -> nil
            end

          _ ->
            nil
        end
      end)
      |> Enum.reject(&is_nil/1)

    port =
      if Enum.empty?(existing_ports) do
        @base_port
      else
        Enum.max(existing_ports) + 1
      end

    File.write!(Path.join(agent_dir, ".env"), "export PORT=#{port}\nexport PHX_HOST=localhost\n")
    port
  end

  defp trust_mise(dir) do
    if System.find_executable("mise") do
      IO.puts("🔐 Trusting Mise Configuration...")
      System.cmd("mise", ["trust", dir])
    end
  end

  defp symlink_assets(dir) do
    IO.puts("🔗 Symlinking Dependencies (Space Saver)...")
    root = File.cwd!()

    # Heavy folders to share
    targets = [
      "node_modules",
      "priv/plts"
    ]

    for target <- targets do
      source = Path.join(root, target)
      destination = Path.join(dir, target)

      if File.exists?(source) do
        # Remove any existing copy in the worktree
        File.rm_rf!(destination)
        File.ln_s!(source, destination)
      end
    end
  end

  defp ensure_ignored do
    gitignore = ".gitignore"

    if File.exists?(gitignore) do
      content = File.read!(gitignore)

      unless String.contains?(content, "#{@agents_root}/") do
        File.write!(gitignore, "\n# Multi-Agent Sandboxes\n#{@agents_root}/\n", [:append])
      end
    end
  end
end

File: scripts/agent-setup.py
#!/usr/bin/env python3

"""
Automates the isolation of an AI agent using Git Worktrees.
One Agent = One Branch = One Directory = One Port.

Usage: ./agent-setup.py <agent_name>
"""

import sys
import subprocess
import os
import re
from pathlib import Path
from typing import List, Optional

# Constants
AGENTS_ROOT = "agents"
BASE_PORT = 4001

# Color codes
CYAN = "\033[1;36m"
GREEN = "\033[1;32m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"


def run_command(cmd: List[str], check: bool = True) -> subprocess.CompletedProcess:
    """Run a shell command and return the result."""
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, check=False)
        if check and result.returncode != 0:
            raise RuntimeError(f"Command failed: {' '.join(cmd)}\n{result.stderr}")
        return result
    except FileNotFoundError as e:
        raise RuntimeError(f"Command not found: {cmd[0]}") from e


def setup_worktree(agent_dir: str, branch: str) -> None:
    """Create git branch and worktree."""
    print("📦 Creating Git Worktree...")

    # Ensure agent root exists
    Path(AGENTS_ROOT).mkdir(parents=True, exist_ok=True)

    # Add to .gitignore if not already there
    gitignore_path = Path(".gitignore")
    if gitignore_path.exists():
        content = gitignore_path.read_text()
        if f"{AGENTS_ROOT}/" not in content:
            with open(gitignore_path, "a") as f:
                f.write(f"\n# Multi-Agent Sandboxes\n{AGENTS_ROOT}/\n")

    # Create branch if it doesn't exist
    result = run_command(["git", "rev-parse", "--verify", branch], check=False)
    if result.returncode != 0:
        run_command(["git", "branch", branch, "main"])
    else:
        print(f"   (Using existing branch: {branch})")

    # Add worktree
    result = run_command(["git", "worktree", "add", agent_dir, branch], check=False)
    if result.returncode != 0:
        raise RuntimeError(f"Could not create worktree: {result.stderr}")


def assign_port(agent_dir: str) -> int:
    """Assign a unique port by scanning existing agents."""
    print("🔢 Assigning Parallel Port...")

    existing_ports = []
    agents_path = Path(AGENTS_ROOT)

    # Scan existing agents for ports
    if agents_path.exists():
        for env_file in agents_path.glob("*/.env"):
            try:
                content = env_file.read_text()
                match = re.search(r"PORT=(\d+)", content)
                if match:
                    existing_ports.append(int(match.group(1)))
            except Exception:
                pass

    port = BASE_PORT if not existing_ports else max(existing_ports) + 1

    # Create .env file
    env_path = Path(agent_dir) / ".env"
    env_path.write_text(f"export PORT={port}\nexport PHX_HOST=localhost\n")

    return port


def trust_mise(agent_dir: str) -> None:
    """Trust mise configuration if available."""
    result = run_command(["which", "mise"], check=False)
    if result.returncode == 0:
        print("🔐 Trusting Mise Configuration...")
        run_command(["mise", "trust", agent_dir])


def symlink_assets(agent_dir: str) -> None:
    """Create symlinks to heavy dependencies."""
    print("🔗 Symlinking Dependencies (Space Saver)...")

    root_dir = Path.cwd()
    targets = ["node_modules", "priv/plts"]

    for target in targets:
        source = root_dir / target
        destination = Path(agent_dir) / target

        if source.exists():
            # Remove any existing copy in the worktree
            if destination.exists() or destination.is_symlink():
                if destination.is_dir() and not destination.is_symlink():
                    import shutil
                    shutil.rmtree(destination)
                else:
                    destination.unlink()

            # Create symlink
            destination.symlink_to(source)


def main() -> None:
    """Main entry point."""
    if len(sys.argv) < 2:
        print("Usage: ./agent-setup.py <agent_name>", file=sys.stderr)
        sys.exit(1)

    agent_name = sys.argv[1]
    agent_dir = f"{AGENTS_ROOT}/{agent_name}"
    branch = f"agent/{agent_name}"

    print(f"\n{CYAN}--- 🤖 Setting up Sandbox for Agent: {agent_name} ---{RESET}")

    try:
        # 1. Create Branch & Worktree
        setup_worktree(agent_dir, branch)

        # 2. Assign Unique Port
        port = assign_port(agent_dir)

        # 3. Trust Mise configuration
        trust_mise(agent_dir)

        # 4. Symlink Heavy Dependencies
        symlink_assets(agent_dir)

        # 5. Final Verification
        print(f"\n{GREEN}✅ Sandbox Ready at: {agent_dir}{RESET}")
        print(f"{BOLD}Agent Branch:{RESET} {branch}")
        print(f"{BOLD}Parallel Port:{RESET} {port}")
        print(
            f"{DIM}(To start the agent's server, run: PORT={port} iex -S mix phx.server in that directory){RESET}\n"
        )

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
